cmd/clusterctl/client/tree/util.go

```diff
@@ -90,14 +90,32 @@ func objToSetter(obj client.Object) conditions.Setter {
 	return setter
 }
 
-// VirtualObject return a new virtual object.
-func VirtualObject(namespace, kind, name string, uid *string) *unstructured.Unstructured {
+// Deprecated: Use VirtualObjectWithUID instead. This function will be dropped in a future release.
+func VirtualObject(namespace, kind, name string) *unstructured.Unstructured {
 	gk := "virtual.cluster.x-k8s.io/v1beta1"
-	// Allow setting a UID for virtual objects so they won't collide if they have the same kind and name.
-	objectUid := fmt.Sprintf("%s, Kind=%s, %s/%s", gk, kind, namespace, name)
-	if uid != nil {
-		objectUid = *uid
+	return &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": gk,
+			"kind":       kind,
+			"metadata": map[string]interface{}{
+				"namespace": namespace,
+				"name":      name,
+				"annotations": map[string]interface{}{
+					VirtualObjectAnnotation: "True",
+				},
+				"uid": fmt.Sprintf("%s, %s/%s", gk, namespace, name),
+			},
+		},
+	}
+}
+
+// VirtualObjectWithIdentifier return a new virtual object with an optional identifier to allow reusing the same name without UID collisions.
+func VirtualObjectWithIdentifier(namespace, kind, name, identifier string) *unstructured.Unstructured {
```

Comment:
I find this kind of confusing, given that an object in Kubernetes should be unique by kind/namespace/name (and UID is an internal detail)
Instead, I think we should create nodes with different names and apply the `ObjectMetaNameAnnotation` if we want that the name that should be used for the object in the presentation layer is different from the actual name